tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp

```diff
@@ -2,74 +2,66 @@
 
-    // Validate dangling story
-    STATIC_ASSERT(
-        same_as<decltype(for_each(borrowed<false>{}, identity{})), for_each_result<ranges::dangling, identity>>);
-    STATIC_ASSERT(same_as<decltype(for_each(borrowed<true>{}, identity{})), for_each_result<int*, identity>>);
+// Validate dangling story
+STATIC_ASSERT(same_as<decltype(ranges::for_each(borrowed<false>{}, identity{})),
+    ranges::for_each_result<ranges::dangling, identity>>);
+STATIC_ASSERT(
+    same_as<decltype(ranges::for_each(borrowed<true>{}, identity{})), ranges::for_each_result<int*, identity>>);
 
-    R const values = {{{0, 42}, {2, 42}, {4, 42}}};
-    auto incr      = [](auto& y) { ++y; };
+struct instantiator {
+    static constexpr P expected[3] = {{1, 42}, {3, 42}, {5, 42}};
 
-    {
-        auto pairs  = values;
-        auto result = for_each(basic_borrowed_range{pairs}, incr, get_first);
-        STATIC_ASSERT(same_as<decltype(result), for_each_result<iterator_t<basic_borrowed_range<P>>, decltype(incr)>>);
-        assert(result.in == basic_borrowed_range{pairs}.end());
-        int some_value = 1729;
-        result.fun(some_value);
-        assert(some_value == 1730);
-        R const expected = {{{1, 42}, {3, 42}, {5, 42}}};
-        assert(ranges::equal(pairs, expected));
-    }
-    {
-        auto pairs = values;
-        basic_borrowed_range wrapped_pairs{pairs};
-        auto result = for_each(wrapped_pairs.begin(), wrapped_pairs.end(), incr, get_second);
-        STATIC_ASSERT(same_as<decltype(result), for_each_result<iterator_t<basic_borrowed_range<P>>, decltype(incr)>>);
-        assert(result.in == wrapped_pairs.end());
-        int some_value = 1729;
-        result.fun(some_value);
-        assert(some_value == 1730);
-        R const expected = {{{0, 43}, {2, 43}, {4, 43}}};
-        assert(ranges::equal(pairs, expected));
-    }
-}
+    template <ranges::input_range ReadWrite>
+    static constexpr void call() {
+        using ranges::for_each, ranges::for_each_result, ranges::iterator_t;
 
-int main() {
-    STATIC_ASSERT((smoke_test(), true));
-    smoke_test();
-}
+        { // Validate iterator + sentinel overload
+            P input[3] = {{0, 42}, {2, 42}, {4, 42}};
+            ReadWrite wrapped_input{input};
 
-struct instantiator {
-    template <class In>
-    static void call(In&& in = {}) {
-        using I   = ranges::iterator_t<In>;
-        using Fun = void (*)(std::iter_common_reference_t<I>);
-        ranges::for_each(in, Fun{});
-        ranges::for_each(ranges::begin(in), ranges::end(in), Fun{});
+            auto result = for_each(wrapped_input.begin(), wrapped_input.end(), incr, get_first);
+            STATIC_ASSERT(
+                same_as<decltype(result), for_each_result<iterator_t<ReadWrite>, remove_const_t<decltype(incr)>>>);
+            assert(result.in == wrapped_input.end());
+            assert(ranges::equal(expected, input));
+
+            int some_value = 1729;
+            result.fun(some_value);

```

Comment:
I observe that this test verifies that `result.fun` is invocable, but not that it actually contains the state accumulated by being invoked on each element. (An incorrect implementation could return a copy of the original callable object.) No change requested.